### PR TITLE
feat(widnows-installer): Add compilation windows installer build step

### PR DESCRIPTION
**Pull request by borismelnik**
Closes: #2223

### CODE REVIEW

File: status.iss [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+; Icon file
+SetupIconFile=resources\status.ico
+
+; Compression
+Compression=lzma
+SolidCompression=yes
+
+;[Languages] - if needed
+;Name: "english"; MessagesFile: "compiler:Default.isl"; LicenseFile: "License_ENG.txt"
+;Name: "russian"; MessagesFile: "compiler:Languages\Russian.isl"; LicenseFile: "License_RUS.txt"
```
Comment (pinebit) [inline status.iss:37]: why not add translations?
e.g. from Telegram installer:
```
[Languages]
Name: "english"; MessagesFile: "compiler:Default.isl"
Name: "it"; MessagesFile: "compiler:Languages\Italian.isl"
Name: "es"; MessagesFile: "compiler:Languages\Spanish.isl"
Name: "de"; MessagesFile: "compiler:Languages\German.isl"
Name: "nl"; MessagesFile: "compiler:Languages\Dutch.isl"
Name: "pt_BR"; MessagesFile: "compiler:Languages\BrazilianPortuguese.isl"
Name: "ru"; MessagesFile: "compiler:Languages\Russian.isl"
Name: "fr"; MessagesFile: "compiler:Languages\French.isl"
Name: "ua"; MessagesFile: "compiler:Languages\Ukrainian.isl"
```

File: status.iss [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+; Generated from Tools -> Generate GUID
+AppId={{F3E2EDB6-78E8-4539-9C8B-A78F059D8647}}
+
+AppName={#Name}
+AppVersion={#Version}
+AppPublisher={#Publisher}
+AppPublisherURL={#URL}
+AppSupportURL={#URL}
+AppUpdatesURL={#URL}
+
```
Comment (pinebit) [inline status.iss:18]: I would suggest to add
`WizardStyle=modern`
to make it looking more or less like Windows 8/10 UI

File: scripts/windows_build_setup.ps1 [hunk 3]
```diff
@@ -63,7 +63,7 @@ export VCINSTALLDIR="/c/BuildTools/VC"
 
 You might also have to include the following paths in your `$PATH:
 
-export PATH=`"/c/BuildTools/MSBuild/Current/Bin:/c/BuildTools/VC/Tools/MSVC/14.27.29110/bin:`$PATH`"
+export PATH=`"/c/ProgramData/scoop/apps/inno-setup/current:/c/BuildTools/MSBuild/Current/Bin:/c/BuildTools/VC/Tools/MSVC/14.27.29110/bin:`$PATH`"
```
Comment (jakubgs) [inline scripts/windows_build_setup.ps1:66]: This is a bit long, split it into 3 lines for clarity:
```
export PATH=`"/c/BuildTools/MSBuild/Current/Bin:`$PATH`"
export PATH=`"/c/BuildTools/VC/Tools/MSVC/14.27.29110/bin:`$PATH`"
export PATH=`"/c/ProgramData/scoop/apps/inno-setup/current:`$PATH`"
```

**Comment by jakubgs** [comment]
Updated CI role to add Inno Setup to `PATH`, but please ping me when this is merged so we can use a `master` branch commit:
https://github.com/status-im/infra-ci/commit/deb54222

**Comment by michaelsbradleyjr** [comment]
A few minor tweaks were needed, which can be found on the `feat/windows-installer-2` branch:

https://github.com/status-im/status-desktop/compare/feat/windows-installer...feat/windows-installer-2

It was important for Jenkins that `STATUS_CLIENT_ZIP` still be defined, which I think is owing to: https://github.com/status-im/status-jenkins-lib/blob/master/vars/windows.groovy#L28. Could be changed to `STATUS_CLIENT_EXE` and then we can drop the redundant variable in `ci/Jenkinsfile.windows` in this repo.

**Review by jakubgs — COMMENTED** [review]
Wouldn't it make more sense to split the build into several targets?
Meaning, separately build the client itself, and then separately zip it up or put into the installer:
```Makefile
STATUS_CLIENT_EXE ?= Status.exe
STATUS_CLIENT_ZIP ?= Status.zip
STATUS_CLIENT_INSTALLER_EXE ?= Status_Installer.exe

$(STATUS_CLIENT_EXE):
  build_client_exe

$(STATUS_CLIENT_ZIP): $(STATUS_CLIENT_EXE)
  build_client_zip_from_exe

$(STATUS_CLIENT_INSTALLER_EXE): $(STATUS_CLIENT_EXE)
  build_installer_from_zip
```
Wouldn't that be cleaner?

**Comment by michaelsbradleyjr** [comment]
The installer does't make use of the zip, it's essentially being deprecated... unless we want to distribute the Windows builds both in installer-less zip form (what we've been doing) and as a self-contained installer exe. I don't think that's the plan but could be wrong (cc: @iurimatias).

We could split out the step that uses windeployqt to build the folder (not compressed) that ISCC packs up into the installer (ISCC does its own compression when building the installer as `pkg/Status.exe`).

In my `feat/windows-installer-2` brach linked above, I was just trying to help get this PR unstuck. But you're right that this PR is a good opportunity to clean up some things.

For example, assuming we'll only distribute the self-contained installer .exe, then we'll no longer need the [nim_windows_launcher](https://github.com/status-im/status-desktop/blob/master/Makefile#L368-L369) target, nor the [nim_windows_launcher.nim](https://github.com/status-im/status-desktop/blob/master/src/nim_windows_launcher.nim) module, nor need to [copy the executable](https://github.com/status-im/status-desktop/blob/master/Makefile#L382) that `nim_windows_launcher` builds from that module.

**Comment by jakubgs** [comment]
>The installer does't make use of the zip

So? Did my example show the installer being dependant on the ZIP?
```
$(STATUS_CLIENT_INSTALLER_EXE): $(STATUS_CLIENT_EXE)
```
Looks to me like it's dependent on the building of the client binary, not the zip in here.

I only included the zip because windows software often gives options to download both an installer and a zip as alternative.

**Comment by michaelsbradleyjr** [comment]
> I only included the zip because windows software often gives options to download both an installer and a zip as alternative.

Sorry about that, I misunderstood.

I do think it would be good to break this down into smaller targets.

Current targets and stuff about them, from perspective of Windows and changes I proposed in [feat/windows-installer-2](https://github.com/status-im/status-desktop/compare/feat/windows-installer...feat/windows-installer-2) branch:
* `nim_status_client`
  * Depends on targets that build and download dependencies of `bin/nim_status_client.exe`.
  * Builds `bin/nim_status_client.exe` with the Nim compiler.

* `$(STATUS_CLIENT_EXE)`
  * Depends on:
    * `nim_status_client`
    * `nim_windows_launcher`, which is not needed since we're building a proper installer.
    * `$(NIM_WINDOWS_PREBUILT_DLLS)`, which should probably be moved into the dependencies of `nim_status_client`.
  * Sets up a deployable folder structure in `tmp/windows/dist/Status`.
  * Copies dependencies into that folder structure that `windeployqt` won't copy automatically.
  * Runs `windeployqt` against `DOtherSide.dll` to copy in the rest of the dependencies.
  * Signs all the `.exe` and `.dll` in the deployable folder via `scripts/sign-windows-bin.sh`.
  * Runs `ISCC` to package the deployable folder into an installer written to `$(STATUS_CLIENT_EXE)`, which by default is `pkg/Status.exe` relative to root of the repo.

### Questions

A quick web search informed me that `ISCC` has the ability to drive the code signing, and I'm guessing it can sign the built installer `.exe` as well, but I don't know that for certain. So do we want to stick with `scripts/sign-windows-bin.sh` or should we transition to having `ISCC` do the code signing?

Looking at the details above, it seems clear `$(STATUS_CLIENT_EXE)` can be broken up into several make targets. Also, there's refactoring work like getting rid of `nim_windows_launcher`. On the other hand, eliminating `nim_windows_launcher` isn't strictly necessary, i.e. the app installed by the installer will work fine if we keep that launcher .exe as part of the installed app for the time being. So what changes and refactors should be part of the MVP installer for Windows, and what improvements to the build steps for the installer can we split into separate tasks for future PRs?

I think it's important to note these questions only pertain to aspects of building the installer. For an MVP installer, there's other work that remains to be done, e.g. it needs to setup appropriate shortcuts in the Start Menu (or whatever it's called these days) and/or the Windows desktop.

Finally, I don't want to step on @borismelnik's toes, this is his PR. I'm just trying to be helpful and provide some feedback to the good points raised by @jakubgs.

**Comment by michaelsbradleyjr** [comment]
Per @borismelnik's request, I've updated this PR's branch with the changes (1 commit) I made on my `feat/windows-installer-2` branch. I also rebased against `master` before force-pushing.

**Comment by jakubgs** [comment]
>So do we want to stick with `scripts/sign-windows-bin.sh` or should we transition to having ISCC do the code signing?

Pretty sure signing of the client binaries themselves and signing of the installer are two different things.

>So what changes and refactors should be part of the MVP installer for Windows, and what improvements to the build steps for the installer can we split into separate tasks for future PRs?

I don't mind doing it in steps, but it definitely needs refactoring, because we can't just endlessly grow one `make` target.
But I'm fine with doing it in a separate PR, but I would still do the bare minimum here and make the installer a separate step that depends on the currently used target.

File: status.iss [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+Name: "nl"; MessagesFile: "compiler:Languages\Dutch.isl"
+Name: "pt_BR"; MessagesFile: "compiler:Languages\BrazilianPortuguese.isl"
+Name: "ru"; MessagesFile: "compiler:Languages\Russian.isl"
+Name: "fr"; MessagesFile: "compiler:Languages\French.isl"
+Name: "ua"; MessagesFile: "compiler:Languages\Ukrainian.isl"
+
+[Files]
+
+; Path to exe on 
+Source: "Status.exe"; DestDir: "{app}"; Flags: ignoreversion
```
Comment (pinebit) [inline status.iss:53]: `{#ExeName}` instead of "Status.exe"

**Comment by michaelsbradleyjr** [comment]
> Pretty sure signing of the client binaries themselves and signing of the installer are two different things.

Sorry if I wasn't clear: it can *definitely* sign the client binaries themselves. I'm guessing it can sign the installer also, but I'm not 100% sure.

**Comment by jakubgs** [comment]
I see, well, then if it's simpler this way we should look into that.

**Comment by anastasiyaig** [comment]
- [ ]  the app is not shown in Start menu after installation

Screenshot 2021-07-15 at 15 00 46


- [ ] unknown publisher is shown when launching installer

Screenshot 2021-07-15 at 14 55 32

- [ ] the app is not centred when launching for the very first time after install

Screenshot 2021-07-15 at 15 01 57


- [ ] app version is wrong in the list of apps in windows app manager

Screenshot 2021-07-15 at 15 07 43

- [ ] status data folder is not removed when deinstalling the app (/AppData/Local/Status)

**Review by michaelsbradleyjr — CHANGES_REQUESTED** [review]
Great work!

Two things:

(1) Before this PR is merged, I think the installer should be modified so that it offers to create shortcuts on the Windows Desktop and/or in the Windows Start Menu.

(2) I am able to reproduce the `JsonParsingError` crash reported by @anastasiyaig. It happens about 1/3 times when creating an account. It also sometimes happens when quitting the app or logging in with an existing account. This does not happen with a CI build of current `master`. In `feat/windows-installer-2`, I tried rebasing this PR branch against `master`; the CI build for that branch crashes the same way.

There's another crash that sometimes happens and involves `writeLine` and Nim's `io.nim` module. It also doesn't happen with a build of current `master`.

Since this PR doesn't change anything in `src/` or `ui/`, I guessed that maybe it's a problem with the folder where the app is installed: `C:\Program Files (x86)\Status`. I tried copying that folder to `C:\Status`.  When opening `C:\Status\Status.exe` I am unable to reproduce either crash mentioned above.

So the crashes do seem to be related somehow to the `C:\Program Files (x86)\Status` path, but I don't know if the problem has something to do with spaces in the path, a permissions problem, or something else. Also, it's very strange that the crashes only happen sometimes.

Before this PR is merged, we need to figure out the nature of the problem and fix it.

**Comment by anastasiyaig** [comment]
@michaelsbradleyjr regarding point 1 i logged an issue already: https://github.com/status-im/status-desktop/issues/2931

**Comment by anastasiyaig** [comment]
regarding the `JsonParsingError`  it happens to me in 3 attempts of 3 when i create new account selecting any item in the list apart of first one, when the app is installed to defaulted path (`C:\Program Files (x86)\Status`). However, i am using Windows virtual machine with Windows 10 Pro:

```Edition	Windows 10 Pro
Version	Dev
Installed on	‎5/‎20/‎2021
OS build	21354.1
Experience	Windows 10 Feature Experience Pack 321.7601.0.3```

designed for Apple Silicon.

This crash does not happen when i install the app to `C:\Status`.



https://user-images.githubusercontent.com/82375995/125911243-8dbb1409-a430-41d0-a59e-39f8a2a7c5ce.mov

**Comment by michaelsbradleyjr** [comment]
I rebased against current `master` and added a couple of commits, 🤞 the Windows installer builds correctly now in Jenkins.

The installer now gives the option to created a desktop shortcut for Status (checked by default) and a "quick launch" icon (unchecked by default). I don't think "quick launch" is even a thing anymore, so we can consider dropping that option.

Also, when you use Windows *Add or remove programs* interface to uninstall Status, the user data folder `%LOCALAPPDATA%\Status` is automatically deleted. I couldn't find a way to make deletion of that folder optional during uninstall. Maybe someone else knows how to do that — seems like it would be a better behavior than automatically deleting it.

**Comment by anastasiyaig** [comment]
@michaelsbradleyjr nice work! What else i noticed so far:

- [ ] when creating a folder in start menu to be status placed in, it is not created. I was expecting a folder, which i can expand and find the status app in. The name of folder could be custom (set during the installation)

Screenshot 2021-07-19 at 10 38 56

- [ ] when creating a shortcut manually and pinning the app to task bar -> they both are not deleted after deinstalling the app

Screenshot 2021-07-19 at 10 56 47


is it possible to fix this? maybe @borismelnik can assist? We can fix it later, these both are small anyway

**Comment by borismelnik** [comment]
> @michaelsbradleyjr nice work! What else i noticed so far:
> 
> * [ ]  when creating a folder in start menu to be status placed in, it is not created. I was expecting a folder, which i can expand and find the status app in. The name of folder could be custom (set during the installation)
> 
> Screenshot 2021-07-19 at 10 38 56
> 
> * [ ]  when creating a shortcut manually and pinning the app to task bar -> they both are not deleted after deinstalling the app (described here [[Windows] Deinstalling the .exe app does not remove shortcuts / task bar pins #2929](https://github.com/status-im/status-desktop/issues/2929) )
> 
> Screenshot 2021-07-19 at 10 56 47
> 
> is it possible to fix this? maybe @borismelnik can assist? We can fix it later, these both are small anyway

Im working on it. About Quick search i think its another option. Will do it too.

File: ci/Jenkinsfile.windows [hunk 5]
```diff
@@ -1,4 +1,4 @@
-library 'status-jenkins-lib@v1.2.18'
+library 'status-jenkins-lib@feat/windows-installer'
```
Comment (jakubgs) [inline ci/Jenkinsfile.windows:1]: @borismelnik why did you merge this like this? The PR was never merged:
https://github.com/status-im/status-jenkins-lib/pull/30

And it was never tagged. I'm pretty sure I told you we'll need to merge it and tag it.

File: ci/Jenkinsfile.windows [hunk 6]
```diff
@@ -26,11 +26,11 @@ pipeline {
   environment {
     TARGET = 'windows'
     /* Improve make performance */
-    MAKEFLAGS = '-j4'
+    MAKEFLAGS = '-j4 V=1'
```
Comment (jakubgs) [inline ci/Jenkinsfile.windows:29]: @borismelnik you left a debug change in the `Jenkinsfile` and merged.

File: ci/Jenkinsfile.windows (hunk 5 above)
Comment (jakubgs) [reply ci/Jenkinsfile.windows:1]: Merged it and tagged it as `v1.2.19`.

**Pull request merged**